### PR TITLE
Fix handling of hardening flags with Clang

### PR DESCRIPTION
**Pull request by Ferroin**
##### Summary

- Ensure that `-Werror` is set when checking compiler flags. Clang has an annoying tendency to ignore some flags it doesn’t actually support, this should get it to properly report them as not supported.
- Explicitly check C and C++ compilers for flag support independently of each other. This ensures we don’t break if the user is using mismatched compilers.
- Fix our Clang CI check to use Clang for _all_ the code, not just the C code.

##### Test Plan

CI passes on this PR.

### CODE REVIEW

**Review by thiagoftsm — APPROVED** [review]
We need to wait for CI, but locally I could compile with `gcc (GCC) 13.2.0` and `clang version 17.0.6`. LGTM!

File: CMakeLists.txt [hunk 1]
```diff
@@ -113,85 +113,149 @@ option(ENABLE_LOGS_MANAGEMENT_TESTS "enable logs management tests" True)
...
+                else()
+                        check_cxx_compiler_flag("-fstack-protector" HAVE_STACK_PROTECTOR)
+                        if(HAVE_STACK_PROTECTOR)
+                                set(EXTRA_HARDENING_CXX_FLAGS "${EXTRA_HARDENING_CXX_FLAGS} -fstack-protector")
                         endif()
                 endif()
         endif()
 
         if(NOT ${CMAKE_C_FLAGS} MATCHES "stack-clash-protection")
                 check_c_compiler_flag("-fstack-clash-protection", HAVE_STACK_CLASH_FLAG)
```
Comment (ilyam8) [inline CMakeLists.txt:150]:
```suggestion
                check_c_compiler_flag("-fstack-clash-protection" HAVE_STACK_CLASH_FLAG)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -113,85 +113,149 @@ option(ENABLE_LOGS_MANAGEMENT_TESTS "enable logs management tests" True)
...
                 if(HAVE_STACK_CLASH_FLAG)
-                        set(EXTRA_HARDENING_FLAGS "${EXTRA_HARDENING_FLAGS} -fstack-clash-protection")
+                        set(EXTRA_HARDENING_C_FLAGS "${EXTRA_HARDENING_C_FLAGS} -fstack-clash-protection")
+                endif()
+        endif()
+
+        if(NOT ${CMAKE_CXX_FLAGS} MATCHES "stack-clash-protection")
+                check_cxx_compiler_flag("-fstack-clash-protection", HAVE_STACK_CLASH_FLAG)
```
Comment (ilyam8) [inline CMakeLists.txt:157]:
```suggestion
                check_cxx_compiler_flag("-fstack-clash-protection" HAVE_STACK_CLASH_FLAG)
```

**Comment by ilyam8** [comment]
@Ferroin out of curiosity I asked Bard if we can simplify the logic using [foreach](https://cmake.org/cmake/help/latest/command/foreach.html), it said yes and gave this example


```c
foreach(LANGUAGE c cxx)
  if(NOT ${CMAKE_${LANGUAGE}_FLAGS} MATCHES "function-sections")
    check_${LANGUAGE}_compiler_flag("-ffunction-sections" HAVE_FUNCTION_SECTIONS)
    if(HAVE_FUNCTION_SECTIONS)
      set(EXTRA_${LANGUAGE}_OPT_FLAGS "${EXTRA_${LANGUAGE}_OPT_FLAGS} -ffunction-sections")
    endif()
  endif()
endforeach()
```

I haven't tested this, just sharing in case you want to try using foreach.

**Comment by ilyam8** [comment]
>  but feel that that should be done in it’s own PR s

As it is now it is error-prone because we duplicate code. I can't track if c and cxx cases are 100% equal (at least visually).

**Comment by ilyam8** [comment]
> Ensure that -Werror is set when checking compiler flags.

This is a problem with gcc: with this option, all tests except FORTIFY_SOURCE fail.

**Comment by Ferroin** [comment]
Rebased to pick up latest changes in `master`.

**Review by ilyam8 — APPROVED** [review]

**Review by thiagoftsm — DISMISSED** [review]
PR is compiling with `gcc (GCC) 13.2.0` and `clang version 17.0.6`.

**Pull request merged**